### PR TITLE
add partial index predicate to SQLiteDialect.get_indexes() result

### LINKED ISSUE #1112: sqlite_where clause gets dropped during migration
**Describe the bug**
In an SQLite database, for an index with `sqlite_where=...` in its definition, the WHERE clause is added when the table is created initially, but then dropped when more indexes are added to the table in later migrations.

**Expected behavior**
I expected that the WHERE clause restricting the index to a certain set of rows is kept during later migrations that do not affect the index.

**To Reproduce**
```py
from alembic import op
import sqlalchemy as sa

# revision identifiers, used by Alembic.
revision = "8b5fc2c6dabe"
down_revision = "7b5fc2c6dabe"
branch_labels = None
depends_on = None


def upgrade():
    op.create_table(
        "emails",
        sa.Column("user_id", sa.BigInteger().with_variant(sa.INTEGER(), "sqlite"), nullable=False),
        sa.Column("primary", sa.Boolean(), nullable=False),
        sa.Column("id", sa.BigInteger().with_variant(sa.INTEGER(), "sqlite"), nullable=False),
    )

    with op.batch_alter_table("emails", schema=None) as batch_op:
        batch_op.create_index(batch_op.f("ix_emails_user_id"), ["user_id"], unique=False)
        batch_op.create_index("single_primary_address", ["user_id"], unique=True,
            postgresql_where=sa.text('"primary"'), sqlite_where=sa.text('"primary"'),
        )

    # the code below recreates the single_primary_address constraint *without* the sqlite_where
    with op.batch_alter_table("emails", schema=None) as batch_op:
        batch_op.create_unique_constraint(batch_op.f("uq_emails_user_id_id"), ["user_id", "id"])


def downgrade():
    pass

```

**Error**

```
2022-11-08T16:38:00+0900 INFO Context impl SQLiteImpl.
2022-11-08T16:38:00+0900 INFO Will assume non-transactional DDL.
2022-11-08T16:38:00+0900 INFO PRAGMA main.table_info("alembic_version")
2022-11-08T16:38:00+0900 INFO [raw sql] ()
FROM alembic_version
2022-11-08T16:38:00+0900 INFO SELECT alembic_version.version_num 
FROM alembic_version
2022-11-08T16:38:00+0900 INFO [generated in 0.00009s] ()
2022-11-08T16:38:00+0900 INFO BEGIN (implicit)
2022-11-08T16:38:00+0900 INFO Running upgrade 7b5fc2c6dabe -> 8b5fc2c6dabe
2022-11-08T16:38:00+0900 INFO 
CREATE TABLE emails (
	user_id INTEGER NOT NULL, 
	"primary" BOOLEAN NOT NULL, 
	id INTEGER NOT NULL
)
2022-11-08T16:38:00+0900 INFO [no key 0.00008s] ()
2022-11-08T16:38:00+0900 INFO CREATE INDEX ix_emails_user_id ON emails (user_id)
2022-11-08T16:38:00+0900 INFO [no key 0.00036s] ()
2022-11-08T16:38:00+0900 INFO CREATE UNIQUE INDEX single_primary_address ON emails (user_id) WHERE "primary"
2022-11-08T16:38:00+0900 INFO [no key 0.00022s] ()
2022-11-08T16:38:00+0900 INFO PRAGMA main.table_xinfo("emails")
2022-11-08T16:38:00+0900 INFO [raw sql] ()
2022-11-08T16:38:00+0900 INFO SELECT sql FROM  (SELECT * FROM sqlite_master UNION ALL   SELECT * FROM sqlite_temp_master) WHERE name = ? AND type = 'table'
2022-11-08T16:38:00+0900 INFO [raw sql] ('emails',)
2022-11-08T16:38:00+0900 INFO PRAGMA main.foreign_key_list("emails")
2022-11-08T16:38:00+0900 INFO [raw sql] ()
2022-11-08T16:38:00+0900 INFO PRAGMA temp.foreign_key_list("emails")
2022-11-08T16:38:00+0900 INFO [raw sql] ()
2022-11-08T16:38:00+0900 INFO SELECT sql FROM  (SELECT * FROM sqlite_master UNION ALL   SELECT * FROM sqlite_temp_master) WHERE name = ? AND type = 'table'
2022-11-08T16:38:00+0900 INFO [raw sql] ('emails',)
2022-11-08T16:38:00+0900 INFO PRAGMA main.index_list("emails")
2022-11-08T16:38:00+0900 INFO [raw sql] ()
2022-11-08T16:38:00+0900 INFO PRAGMA main.index_info("single_primary_address")
2022-11-08T16:38:00+0900 INFO [raw sql] ()
2022-11-08T16:38:00+0900 INFO PRAGMA main.index_info("ix_emails_user_id")
2022-11-08T16:38:00+0900 INFO [raw sql] ()
2022-11-08T16:38:00+0900 INFO PRAGMA main.index_list("emails")
2022-11-08T16:38:00+0900 INFO [raw sql] ()
2022-11-08T16:38:00+0900 INFO PRAGMA main.index_info("single_primary_address")
2022-11-08T16:38:00+0900 INFO [raw sql] ()
2022-11-08T16:38:00+0900 INFO PRAGMA main.index_info("ix_emails_user_id")
2022-11-08T16:38:00+0900 INFO [raw sql] ()
2022-11-08T16:38:00+0900 INFO SELECT sql FROM  (SELECT * FROM sqlite_master UNION ALL   SELECT * FROM sqlite_temp_master) WHERE name = ? AND type = 'table'
2022-11-08T16:38:00+0900 INFO [raw sql] ('emails',)
2022-11-08T16:38:00+0900 INFO 
CREATE TABLE _alembic_tmp_emails (
	user_id INTEGER NOT NULL, 
	"primary" BOOLEAN NOT NULL, 
	id INTEGER NOT NULL, 
	CONSTRAINT uq_emails_user_id_id UNIQUE (user_id, id)
)
2022-11-08T16:38:00+0900 INFO [no key 0.00016s] ()
2022-11-08T16:38:00+0900 INFO INSERT INTO _alembic_tmp_emails (user_id, "primary", id) SELECT emails.user_id, emails."primary", emails.id 
FROM emails
2022-11-08T16:38:00+0900 INFO [generated in 0.00016s] ()
2022-11-08T16:38:00+0900 INFO 
DROP TABLE emails
2022-11-08T16:38:00+0900 INFO [no key 0.00008s] ()
2022-11-08T16:38:00+0900 INFO ALTER TABLE _alembic_tmp_emails RENAME TO emails
2022-11-08T16:38:00+0900 INFO [no key 0.00009s] ()
2022-11-08T16:38:00+0900 INFO CREATE UNIQUE INDEX single_primary_address ON emails (user_id)
2022-11-08T16:38:00+0900 INFO [no key 0.00011s] ()
2022-11-08T16:38:00+0900 INFO CREATE INDEX ix_emails_user_id ON emails (user_id)
2022-11-08T16:38:00+0900 INFO [no key 0.00007s] ()
2022-11-08T16:38:00+0900 INFO UPDATE alembic_version SET version_num='8b5fc2c6dabe' WHERE alembic_version.version_num = '7b5fc2c6dabe'
2022-11-08T16:38:00+0900 INFO [generated in 0.00020s] ()
2022-11-08T16:38:00+0900 INFO COMMIT
```

**Note:** There is *no* `WHERE "primary"` when the second batch operation is running.

**Versions.**
 - OS: openSUSE Linux
 - Python: 3.8.15
 - Alembic: 1.7.7 (checked with 1.8.1 as well, same behavior)
 - SQLAlchemy: 1.4.32
 - Database: SQLite 3.39.3
 - DBAPI: ?

**Additional context**

**Have a nice day!**

### LINKED ISSUE #8804: SQLite reflection drops the WHERE clause of partial indexes
### Describe the bug

`SQLiteDialect.get_indexes()` does not distinguish between full and partial indexes. If an index is partial, that information is not communicated to the caller of `get_indexes()` and this may cause, say, alembic to recreate a partial index as a full index during a migration, cf. https://github.com/sqlalchemy/alembic/issues/1112

In https://github.com/sqlalchemy/sqlalchemy/blob/9237bf15e612ba82555444751bd69dc2a831e7f4/lib/sqlalchemy/dialects/sqlite/base.py#L2647-L2654 we should treat the case `row[4] == 1` separately (see [`index_list()`](https://www.sqlite.org/pragma.html#pragma_index_list)) and then probably need to do some extra work like parsing the string in `sqlite_master.sql` for that particular index in order to get the actual filter predicate.

### To Reproduce

```python
    def test_reflect_partial_indexes(self):
        with testing.db.begin() as conn:
            conn.exec_driver_sql(
                "create table foo_with_partial_index (x integer, y integer)"
            )
            conn.exec_driver_sql(
                "create unique index ix_partial on foo_with_partial_index (x) where y > 10"
            )
            conn.exec_driver_sql(
                "create unique index ix_no_partial on foo_with_partial_index (x)"
            )

            inspector = inspect(conn)
            eq_(
                inspector.get_indexes("foo_with_partial_index"),
                [
                    # the two indexes are indistinguishable
                    {
                        "unique": 1,
                        "name": "ix_no_partial",
                        "column_names": ["x"],
                    },
                    {
                        "unique": 1,
                        "name": "ix_partial",
                        "column_names": ["x"],
                    },
                ],
            )
```

### Error

(no error)

### Versions

- OS: openSUSE Linux
- Python: 3.8.15
- SQLAlchemy: current main (9237bf15e612ba82555444751bd69dc2a831e7f4)
- Database: SQLite
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): ?

### Additional context

Probably this is similar to the non-handling of expression-based indexes, which are also not handled: https://github.com/sqlalchemy/sqlalchemy/blob/9237bf15e612ba82555444751bd69dc2a831e7f4/lib/sqlalchemy/dialects/sqlite/base.py#L2663-L2669

### PR DESCRIPTION
**Pull request by tgpfeiffer**
This fixes #8804 by populating the `dialect_options`'s `sqlite_where` field inside `SQLiteDialect.get_indexes()` for an index where the `index_list()` pragma indicates that it is a partial index.

### Description
For an index where the fifth column of the corresponding row in the `index_list()` return value is true (see https://www.sqlite.org/pragma.html#pragma_index_list), we get the exact SQL used to create that index from the `sqlite_master` table and look for the `WHERE` keyword, then use the string after that as the `sqlite_where` predicate.

Remark: I use a simple regular expression scanning for the string `)\s+where\s+` (case-insensitive) to find the predicate. As written in a comment in the code, this can easily fail if the definition of the index contains a matching string such as
```sql
CREATE INDEX i ON t (col || ') where') WHERE col <> ''         -- case 1
CREATE INDEX i ON t (col)              WHERE col <> ') where'  -- case 2
```
but case 1 can be neglected because `get_indexes()` currently does not support expression-based indexes, and case 2 can be neglected because the regex finds the first match.

@CaselIT suggested in https://github.com/sqlalchemy/sqlalchemy/issues/8804#issuecomment-1314899834 to attack expression-based indexes as well, but to be honest I am not sure what is a good way to parse the `CREATE INDEX` statement with reasonable effort in that case, so I didn't work on it. (Is there a parser somewhere in the codebase already?)

I have verified that in my local environment, the case I reported in https://github.com/sqlalchemy/alembic/issues/1112#issue-1439662864 works well, so the partial index is kept across migrations and this PR also fixes https://github.com/sqlalchemy/alembic/issues/1112.

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [X] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

### CODE REVIEW

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 0bd27506c49e9297f9cc156f45341298b811e833 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change 0bd27506c49e9297f9cc156f45341298b811e833: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4200

**Comment by tgpfeiffer** [comment]
I think I addressed your comments @CaselIT, could you take another look please? Thank you!

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 56f16e30b494d6e58664ea4d66aa07c234398c54 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset 56f16e30b494d6e58664ea4d66aa07c234398c54 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4200

File: lib/sqlalchemy/dialects/sqlite/base.py [hunk 1]
```diff
@@ -2651,7 +2667,38 @@ def get_indexes(self, connection, table_name, schema=None, **kw):
...
+
+            # check partial indexes
+            if row[4]:
+                s = (
+                    "SELECT sql FROM %(schema)ssqlite_master "
+                    "WHERE name = ? "
+                    "AND type = 'index'" % {"schema": schema_expr}
+                )
+                rs = connection.exec_driver_sql(s, (row[1],))
+                try:
```
Comment (sqla-tester) [inline lib/sqlalchemy/dialects/sqlite/base.py:2687]: **Federico Caselli** (CaselIT) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4200

File: lib/sqlalchemy/dialects/sqlite/base.py [hunk 2]
```diff
@@ -2651,7 +2667,38 @@ def get_indexes(self, connection, table_name, schema=None, **kw):
...
+                    predicate = partial_pred_re.search(index_sql).group(1)
+                    indexes[-1]["dialect_options"]["sqlite_where"] = text(
+                        predicate
+                    )
+                except Exception:
+                    util.warn(
+                        "Failed to look up filter predicate of "
+                        "partial index %s" % row[1]
+                    )
+                    indexes.pop()
```
Comment (sqla-tester) [inline lib/sqlalchemy/dialects/sqlite/base.py:2701]: **Federico Caselli** (CaselIT) wrote:

this is a change compared to how the dialect works now (since it currently returns partial indexes without informations instead of ignoring them if it cannot parse them)

but since it's in an edge case that should never happen, at least when used with sqlite, I think it's fine

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4200

File: test/dialect/test_sqlite.py [hunk 3]
```diff
@@ -2297,10 +2299,63 @@ def test_create_index_with_schema(self):
...
             ],
         )
 
+    def test_reflect_partial_indexes(self):
+        with testing.db.begin() as conn:
+            conn.exec_driver_sql(
+                "create table foo_with_partial_index (x integer, y integer)"
+            )
+            conn.exec_driver_sql(
+                "create unique index ix_partial on "
```
Comment (sqla-tester) [inline test/dialect/test_sqlite.py:2313]: **Federico Caselli** (CaselIT) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4200

**Review by sqla-tester — COMMENTED** [review]
**Federico Caselli** (CaselIT) wrote:

looks ok to me now. Thanks for updating the change

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4200

**Comment by tgpfeiffer** [comment]
Sorry I am not familiar with the review process, what is the next step to get this merged, looks like I need another approval?

On a related note, I would also like to see this change in the next 1.4.x bugfix release, can I do anything to help with the backporting? (I already noticed there will be a merge conflict, so it needs some manual work.)

**Comment by CaselIT** [comment]
> Sorry I am not familiar with the review process, what is the next step to get this merged, looks like I need another approval?

Mostly wait, Mike will get to this when he has time

> On a related note, I would also like to see this change in the next 1.4.x bugfix release, can I do anything to help with the backporting? (I already noticed there will be a merge conflict, so it needs some manual work.)

I placed it on the 2 milestone, but it seems contained enough to be backported to 1.4. let's wait for mike opinion on this

File: lib/sqlalchemy/dialects/sqlite/base.py [hunk 4]
```diff
@@ -2651,7 +2667,38 @@ def get_indexes(self, connection, table_name, schema=None, **kw):
...
+                try:
+                    # in theory the code below shouldn't happen because
+                    # we know this is a partial index, so the definition
+                    # sql should be there and match the regex
+                    index_sql = rs.scalar()
+                    predicate = partial_pred_re.search(index_sql).group(1)
+                    indexes[-1]["dialect_options"]["sqlite_where"] = text(
+                        predicate
+                    )
+                except Exception:
```
Comment (sqla-tester) [inline lib/sqlalchemy/dialects/sqlite/base.py:2696]: **mike bayer** (zzzeek) wrote:

what Exception are we catching here?   is this just if the partial_pred_re is None?  Let's just assign it up front and do "if predicate_match is not None"

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4200

File: lib/sqlalchemy/dialects/sqlite/base.py (hunk 2 above)
Comment (sqla-tester) [inline lib/sqlalchemy/dialects/sqlite/base.py:2701]: **mike bayer** (zzzeek) wrote:

this would be much safer to backport to 1.4 if we didnt actually do this indexes.pop() here, and return the same output we would have gotten previously.  We can leave the new warning in place, which should be sufficient.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4200

File: test/dialect/test_sqlite.py [hunk 5]
```diff
@@ -4,6 +4,7 @@
 import datetime
 import json
 import os
+from unittest.mock import ANY
```
Comment (sqla-tester) [inline test/dialect/test_sqlite.py:7]: **mike bayer** (zzzeek) wrote:

for backporting this has to be "from testing.mock import ANY"

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4200

File: test/dialect/test_sqlite.py [hunk 6]
```diff
@@ -2297,10 +2299,63 @@ def test_create_index_with_schema(self):
                     "unique": 0,
                     "name": "ix_main_l_bar",
                     "column_names": ["bar"],
+                    "dialect_options": {},
                 }
             ],
         )
 
+    def test_reflect_partial_indexes(self):
```
Comment (sqla-tester) [inline test/dialect/test_sqlite.py:2307]: **mike bayer** (zzzeek) wrote:

use the connection fixture

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4200

File: test/dialect/test_sqlite.py [hunk 7]
```diff
@@ -2297,10 +2299,63 @@ def test_create_index_with_schema(self):
...
+        with testing.db.begin() as conn:
+            conn.exec_driver_sql(
+                "create table foo_with_partial_index (x integer, y integer)"
```
Comment (sqla-tester) [inline test/dialect/test_sqlite.py:2310]: **mike bayer** (zzzeek) wrote:

do all of these individually using a @testing.combinations() block

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4200

**Review by sqla-tester — COMMENTED** [review]
**mike bayer** (zzzeek) wrote:

hi thanks for this!   this is pretty good. I'm assuming we want this in 1.4 as 2.0 final is still some months away.    Let's narrow the scope of the new logic, and there are also changes for the test suite; in the latter case, we can do that part if you dont know our test suite conventions.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4200

**Comment by tgpfeiffer** [comment]
(Not sure why some reviewer comments show up in github and others in gerrit only... :thinking:)

I addressed three of your comments (fix mock import, remove `indexes.pop()`, use an `is None` for the regex). As for the `@testing.combinations()` block, as @CaselIT wrote in https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4200/2/test/dialect/test_sqlite.py#2310 I wrote it that way to test that we got the list right for multiple indexes. About the connection fixture @CaselIT wrote "Done" in https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4200/comment/de64c240_86533cb7/ but I am not sure *where* it is done :sweat_smile: 

Please take another look and let me know if I should fix anything else.

**Comment by zzzeek** [comment]
> (Not sure why some reviewer comments show up in github and others in gerrit only... thinking)

we have a mirror between the gerrit review and this PR.   they *should* be bidrectional at the comments level.

> 
> I addressed three of your comments (fix mock import, remove `indexes.pop()`, use an `is None` for the regex). As for the `@testing.combinations()` block, as @CaselIT wrote in https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4200/2/test/dialect/test_sqlite.py#2310 I wrote it that way to test that we got the list right for multiple indexes. About the connection fixture @CaselIT wrote "Done" in https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4200/comment/de64c240_86533cb7/ but I am not sure _where_ it is done sweat_smile

that would replace the `engine.begin()` and would look like this:

```py

  def test_reflect_partial_indexes(self, connection):
        connection.exec_driver_sql(...)

```

the thing we are using there is [pytest fixtures](https://docs.pytest.org/en/7.2.x/how-to/fixtures.html#how-to-fixtures).

**Comment by CaselIT** [comment]
yes, it was a mistake on my part that done comment

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 44cd132517232bde50264c9516c4347c5b297e10 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by tgpfeiffer** [comment]
Pushed another commit making use of the `connection` feature. I think I addressed all your comments now.

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 539dfcb372360911b69aed2a804698bb1a2220b1 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset 539dfcb372360911b69aed2a804698bb1a2220b1 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4200

**Comment by tgpfeiffer** [comment]
Sorry to bug you again, could you give this another round of review, please? Thank you!

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 539dfcb372360911b69aed2a804698bb1a2220b1 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset 539dfcb372360911b69aed2a804698bb1a2220b1 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4200

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4200 has been **merged**. Congratulations! :)

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4251 has been **merged**. Congratulations! :)
